### PR TITLE
Rewrite `sum(if())` and `sumIf` to `countIf` in special cases

### PR DESCRIPTION
**Pull request by ucasfl**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):

- Performance Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Rewrite `sumIf()` and `sum(if())` function to `countIf()` function when logically equivalent.

### CODE REVIEW

**Comment by ucasFL** [comment]
I'm confusing about the specialized implementation of countIf, does it have special optimization? Current implementation is worse maybe, because currently, `select countIf(1) from numbers(1000000000)` is too much slower than `select sum(if(1, 1, 0)) from numbers(1000000000)`. @alexey-milovidov

**Comment by amosbird** [comment]
@ucasFL It's because countIf is neither devirtualized nor vectorized.

File: src/Interpreters/RewriteSumIfFunctonVisitor.cpp [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    if (lower_name != "sum" && lower_name != "sumif")
+        return;
+
+    auto & func_arguments = func.arguments->children;
+
+    if (lower_name == "sumif")
+    {
+        /// sumIf(1, cond) -> countIf(cond)
+        const auto * literal = func_arguments[0]->as<ASTLiteral>();
+        if (func_arguments.size() == 2 && literal && literal->value.get<Int64>() == 1)
```
Comment (vdimir) [inline src/Interpreters/RewriteSumIfFunctonVisitor.cpp:32]: Probably type of constant should be UInt64, and better to call `tryGet` or check type explicitly. Here and in code below

File: src/Interpreters/RewriteSumIfFunctonVisitor.cpp [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        {
+            auto new_func = std::make_shared<ASTFunction>();
+            new_func->name = "countIf";
+
+            auto new_arguments = std::make_shared<ASTExpressionList>();
+            new_arguments->children.push_back(func_arguments[1]);
+            new_func->arguments = new_arguments;
+            new_func->children.push_back(new_arguments);
+            new_func->setAlias(func.alias);
+            ast = std::move(new_func);
```
Comment (vdimir) [inline src/Interpreters/RewriteSumIfFunctonVisitor.cpp:42]: This code duplicates three times, maybe we can try to extract it to separate function or rewrite in other way.

**Comment by vdimir** [comment]
Can you add test for new feature with flag enabled?

File: src/Core/Settings.h [hunk 3]
```diff
@@ -413,6 +413,7 @@ class IColumn;
     \
     M(Bool, async_socket_for_remote, true, "Asynchronously read from socket executing remote query", 0) \
     \
+    M(Bool, optimize_rewrite_sum_if_to_count_if, true, "Rewrite sumIf() and sum(if()) function countIf() function when logic equivalen", 0) \
```
[the remark this answers is not in the document]
Comment (stigsb) [reply src/Core/Settings.h:416]: Spelling nit: "logically equivalent"

[the remark this answers is not in the document]
Comment (vdimir) [reply src/Core/Settings.h:416]: > Spelling nit: "logically equivalent"

Thanks

**Comment by ucasFL** [comment]
> Could you add test for new feature with flag enabled?
> Also it would be nice to have performance comparison.

Ok!

File: tests/performance/rewrite_sumIf.xml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+<test>
+    <query>SELECT sumIf(1, 0) FROM numbers(100000000)</query>
```
Comment (vdimir) [inline tests/performance/rewrite_sumIf.xml:2]: Should flag `optimize_rewrite_sum_if_to_count_if` be enabled to compare new behavior with old? Also it is good to add functional test, to ensure that it works correctly.

Comment (ucasFL) [reply tests/performance/rewrite_sumIf.xml:2]: The flag is enabled default, and we can see the difference:
https://clickhouse-test-reports.s3.yandex.net/17041/1f7b9053311febfbbaa17b71cb265e6a9705b117/performance_comparison/report.html

Ok, i will add a functional test.

**Review by vdimir — APPROVED** [review]
Thank your for PR!

I've only added a bit more complex queries to test to ensure that expression aliases still work with rewrite

**Comment by vdimir** [comment]
Seems that failed test not connected to changes:
Integration tests (asan): `test_zookeeper_config/test.py::test_secure_connection`
Functional stateless tests (thread): `01085_max_distributed_connections_http`

**Pull request merged**